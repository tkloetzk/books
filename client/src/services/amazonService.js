import apiConfig from '../config/apiConfig';
import axios from 'axios';

export function getAmazonBookService(isbns) {
  return axios
    .post(apiConfig.amazon, { isbns })
    .then(books => books.data)
    .catch(error => {
      throw error;
    });
}

export function getAmazonBookServiceSingle(isbn) {
  return axios
    .post(apiConfig.amazonV2, { isbn })
    .then(res => res.data)
    .catch(error => {
      throw error;
    });
}
