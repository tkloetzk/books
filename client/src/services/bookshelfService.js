import apiConfig from '../config/apiConfig';
import axios from 'axios';
import sortBooklist from '../util/calculator';

export function getBookshelfService(excludedGenres) {
  return axios
    .post(apiConfig.bookshelf, excludedGenres)
    .then(res => sortBooklist(res.data))
    .catch(error => {
      throw error;
    });
}
export function addBookshelfService(booklist) {
  return axios
    .post(`${apiConfig.bookshelf}/add`, booklist)
    .then(res => console.log(res.data))
    .catch(error => {
      throw error;
    });
}

export function updateBookOnBookshelfService(id, fields) {
  return axios
    .put(`${apiConfig.bookshelf}/update/${id}`, fields)
    .then(res => console.log(res.data))
    .catch(error => {
      throw error;
    });
}
