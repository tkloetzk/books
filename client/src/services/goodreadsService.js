import apiConfig from '../config/apiConfig';
import axios from 'axios';

export function getGoodreadsSingleBooksService(booklist) {
  return axios
    .get(`${apiConfig.goodreads}/${booklist}`)
    .then(booklist => booklist.data)
    .catch(error => {
      throw error;
    });
}
