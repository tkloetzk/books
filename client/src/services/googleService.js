import apiConfig from '../config/apiConfig';
import axios from 'axios';

export function getGoogleBookService(isbn) {
  return axios
    .get(`${apiConfig.google}/${isbn}`)
    .then(book => book.data)
    .catch(error => {
      throw error;
    });
}
