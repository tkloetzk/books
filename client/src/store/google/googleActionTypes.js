export const FETCH_GOOGLE_BOOK_HAS_ERRORED = 'FETCH_GOOGLE_BOOK_HAS_ERRORED';
export const FETCH_GOOGLE_BOOK_IS_LOADING = 'FETCH_GOOGLE_BOOK_IS_LOADING';
export const FETCH_GOOGLE_BOOK_SUCCESS = 'FETCH_GOOGLE_BOOK_SUCCESS';
