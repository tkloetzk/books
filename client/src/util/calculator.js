import get from 'lodash/get';
import forEach from 'lodash/forEach'
import clone from 'lodash/clone'

let meanGoodreadsVotes
let meanAmazonVotes
let totalMean

export default function sortBooklist(originalBooklist) {
  const booklist = clone(originalBooklist);
  calculateMeans(booklist);
  calculateAdjustedRating(booklist);
  booklist.sort((a, b) =>
    a.adjustedRating < b.adjustedRating
      ? 1
      : b.adjustedRating < a.adjustedRating
      ? -1
      : 0
  );
  return booklist;
}

function calculateMeans(booklist) {
  let goodreadsVotes = 0,
    amazonVotes = 0,
    total = 0;

  forEach(booklist, book => {
    goodreadsVotes += get(book, 'goodreadsAverageRating')
    amazonVotes += get(book, 'amazonAverageRating');
    total +=
      get(book, 'goodreadsAverageRating') +
      get(book, 'amazonAverageRating');
  })

  meanGoodreadsVotes = goodreadsVotes / booklist.length;
  meanAmazonVotes = amazonVotes / booklist.length;
  totalMean = total / 2 / booklist.length;

  return booklist;
}

function calculateAdjustedRating(booklist) {
  for (var i = 0; i < booklist.length; i++) {
    var book = booklist[i];
    var adjustedRating =
      (getAdjustedRating(
        get(book, 'goodreadsRatingsCount'),
        get(book, 'goodreadsAverageRating'),
        meanGoodreadsVotes,
        trimmean(getGoodreadsRatingsCountList(booklist))
      ) +
        getAdjustedRating(
          get(book, 'amazonRatingsCount'),
          get(book, 'amazonAverageRating'),
          meanAmazonVotes,
          trimmean(getAmazonRatingsCountList(booklist))
        )) /
      2;
    book.adjustedRating = adjustedRating;
  }
}

function getAdjustedRating(ratingsCount, averageRating, meanVote, minVotes) {
  //double rating = averageRating * 2;
  return (
    (ratingsCount / (ratingsCount + minVotes)) * averageRating +
    (minVotes / (ratingsCount + minVotes)) * meanVote
  );
}

function getGoodreadsRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.goodreadsRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function getAmazonRatingsCountList(booklist) {
  let votes = [];
  for (const book of booklist) {
    votes.push(book.amazonRatingsCount);
  }

  return votes.sort(function(a, b) {
    return a - b;
  });
}

function trimmean(votes) {
  votes.sort(function(a, b) {
    return a - b;
  });

  const removeAmount = (votes.length * 0.25) / 2;
  const roundDownNearedMultipleTwo = Math.floor(
    removeAmount >= 0
      ? (removeAmount / 2) * 2
      : ((removeAmount - 2 + 1) / 2) * 2
  );
  return getMean(
    votes.slice(
      roundDownNearedMultipleTwo,
      votes.length - roundDownNearedMultipleTwo
    )
  );
}

function getMean(array) {
  let sum = 0;
  for (let i = 0; i < array.length; i++) {
    sum += array[i];
  }
  return sum / array.length;
}
