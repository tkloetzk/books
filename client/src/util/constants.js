export const LOADING_STATUSES = {
  initial: 'initial',
  loading: 'loading',
  success: 'success',
  errored: 'errored',
};
