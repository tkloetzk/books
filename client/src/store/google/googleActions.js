import * as types from './googleActionTypes';
import { getGoogleBookService } from '../../services/googleService';
import { LOADING_STATUSES } from '../../util/constants';

export function getGoogleBookFailure(bool) {
  return {
    type: types.FETCH_GOOGLE_BOOK_HAS_ERRORED,
    hasErrored: bool,
  };
}

export function getGoogleBookIsLoading(status) {
  return {
    type: types.FETCH_GOOGLE_BOOK_IS_LOADING,
    isLoading: status,
  };
}

export function getGoogleBookSuccess(book) {
  return {
    type: types.FETCH_GOOGLE_BOOK_SUCCESS,
    book,
  };
}

export function getGoogleBook(isbn) {
  return dispatch => {
    dispatch(getGoogleBookIsLoading(LOADING_STATUSES.loading));
    return getGoogleBookService(isbn)
      .then(resp => {
        dispatch(getGoogleBookIsLoading(LOADING_STATUSES.success));
        dispatch(getGoogleBookSuccess(resp));
      })
      .catch(error => {
        dispatch(getGoogleBookIsLoading(LOADING_STATUSES.errored));
        dispatch(getGoogleBookFailure(true));
      });
  };
}

export function clearGoogleBooksSuccess() {
  return {
    type: types.CLEAR_GOOGLE_BOOKS_SUCCESS,
  };
}

export function clearGoogleBooks() {
  return dispatch => {
    dispatch(clearGoogleBooksSuccess());
  };
}
