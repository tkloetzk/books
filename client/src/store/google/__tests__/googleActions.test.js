import * as actions from '../googleActions';
import * as types from '../googleActionTypes';
import { LOADING_STATUSES } from '../../../util/constants';
import { getGoogleBookService } from '../../../services/googleService';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const createMockStore = configureMockStore([thunk]);

jest.mock('../../../services/googleService');

describe('google actions', () => {
  let book;
  beforeEach(() => {
    book = {
      title: 'Operating Instructions',
      isbn: '9781400079094',
      subtitle: "A Journal Of My Son's First Year",
      description: 'A single mother and writer grappling alone',
      thumbnail:
        'http://books.google.com/books/content?id=MiKz2RreyecC&printsec=frontcover&img=1&zoom=1&edge=curl&source=gbs_api',
      categories: ['Family & Relationships'],
    };
    getGoogleBookService.mockClear();
  });
  it('can set google book hasErrored', () => {
    const actionTrue = actions.getGoogleBookFailure(true);
    expect(actionTrue).toEqual({
      type: types.FETCH_GOOGLE_BOOK_HAS_ERRORED,
      hasErrored: true,
    });

    const actionFalse = actions.getGoogleBookFailure(false);
    expect(actionFalse).toEqual({
      type: types.FETCH_GOOGLE_BOOK_HAS_ERRORED,
      hasErrored: false,
    });
  });
  it('can set google book isLoading', () => {
    const action = actions.getGoogleBookIsLoading(LOADING_STATUSES.initial);
    expect(action).toEqual({
      type: types.FETCH_GOOGLE_BOOK_IS_LOADING,
      isLoading: LOADING_STATUSES.initial,
    });
  });
  it('can set google book success', () => {
    const action = actions.getGoogleBookSuccess(book);
    expect(action).toEqual({
      type: types.FETCH_GOOGLE_BOOK_SUCCESS,
      book,
    });
  });
  it('can clear google books', () => {
    const dispatch = jest.fn();
    actions.clearGoogleBooks()(dispatch);
    expect(dispatch.mock.calls).toMatchSnapshot();
  });
  it('can clear google books success', () => {
    const action = actions.clearGoogleBooksSuccess();
    expect(action).toEqual({
      type: types.CLEAR_GOOGLE_BOOKS_SUCCESS,
    });
  });
  it('can get google book', () => {
    const store = createMockStore();
    const response = {
      title: 'A Curve in the Road',
      isbn: '9781503904453',
      subtitle: '',
      description:
        'Abbie MacIntyre is living the dream in the picturesque Nova Scotia town she calls home. She is a successful surgeon, is married to a handsome cardiologist, and has a model teenage son who is only months away from going off to college. But then one fateful night, everything changes. When a drunk driver hits her car, Abbie is rushed to the hospital. She survives, but the accident forces unimaginable secrets out into the open and plagues Abbie with nightmares so vivid that she starts to question her grip on reality. Her perfect life begins to crack, and those cracks threaten to shatter her world completely. The search for answers will test her strength in every way—as a wife, a career woman, and a mother—but it may also open the door for Abbie to move forward, beyond anger and heartbreak, to find out what she is truly made of. In learning to heal and trust again, she may just find new hope in the spaces left behind.',
      thumbnail:
        'http://books.google.com/books/content?id=SPLJswEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api',
      categories: ['Drunk driving'],
    };
    const expectedResponse = [
      { isLoading: 'loading', type: 'FETCH_GOOGLE_BOOK_IS_LOADING' },
      { isLoading: 'success', type: 'FETCH_GOOGLE_BOOK_IS_LOADING' },
      {
        book: {
          categories: ['Drunk driving'],
          description:
            'Abbie MacIntyre is living the dream in the picturesque Nova Scotia town she calls home. She is a successful surgeon, is married to a handsome cardiologist, and has a model teenage son who is only months away from going off to college. But then one fateful night, everything changes. When a drunk driver hits her car, Abbie is rushed to the hospital. She survives, but the accident forces unimaginable secrets out into the open and plagues Abbie with nightmares so vivid that she starts to question her grip on reality. Her perfect life begins to crack, and those cracks threaten to shatter her world completely. The search for answers will test her strength in every way—as a wife, a career woman, and a mother—but it may also open the door for Abbie to move forward, beyond anger and heartbreak, to find out what she is truly made of. In learning to heal and trust again, she may just find new hope in the spaces left behind.',
          isbn: '9781503904453',
          subtitle: '',
          thumbnail:
            'http://books.google.com/books/content?id=SPLJswEACAAJ&printsec=frontcover&img=1&zoom=1&source=gbs_api',
          title: 'A Curve in the Road',
        },
        type: 'FETCH_GOOGLE_BOOK_SUCCESS',
      },
      {
        error: undefined,
        hasErrored: false,
        type: 'FETCH_GOOGLE_BOOK_HAS_ERRORED',
      },
    ];
    getGoogleBookService.mockResolvedValue(response);

    return store.dispatch(actions.getGoogleBook(book.isbn)).then(() => {
      expect(store.getActions()).toEqual(expectedResponse);
    });
  });
  it('can throw error on get google book', () => {
    const store = createMockStore();
    const response = null;
    const expectedResponse = [
      { isLoading: 'loading', type: 'FETCH_GOOGLE_BOOK_IS_LOADING' },
      { isLoading: 'errored', type: 'FETCH_GOOGLE_BOOK_IS_LOADING' },
      { error: null, hasErrored: true, type: 'FETCH_GOOGLE_BOOK_HAS_ERRORED' },
    ];
    getGoogleBookService.mockRejectedValue(response);

    return store.dispatch(actions.getGoogleBook(book.isbn)).then(() => {
      expect(store.getActions()).toEqual(expectedResponse);
    });
  });
});
