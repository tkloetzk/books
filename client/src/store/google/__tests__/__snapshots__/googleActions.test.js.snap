// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`google actions can clear google books 1`] = `
Array [
  Array [
    Object {
      "type": "CLEAR_GOOGLE_BOOKS_SUCCESS",
    },
  ],
]
`;
